pennylane/_qubit_device.py

```diff
@@ -812,21 +833,52 @@ def statistics(self, circuit: QuantumScript, shot_range=None, bin_size=None):
                 )
 
             elif obs.return_type is Shadow:
-                if len(observables) > 1:
+                if len(measurements) > 1:
                     raise qml.QuantumFunctionError(
                         "Classical shadows cannot be returned in combination"
                         " with other return types"
                     )
                 results.append(self.classical_shadow(obs, circuit))
 
             elif obs.return_type is ShadowExpval:
-                if len(observables) > 1:
+                if len(measurements) > 1:
                     raise qml.QuantumFunctionError(
                         "Classical shadows cannot be returned in combination"
                         " with other return types"
                     )
                 results.append(self.shadow_expval(obs, circuit=circuit))
 
+            elif method := getattr(self, m.method_name, False):
+                results.append(method(obs, shot_range=shot_range, bin_size=bin_size))
+
+            elif isinstance(m, CustomMeasurement):
+                results.append(m.process(tape=circuit, device=self))
+
+            elif isinstance(m, SampleMeasurement) and self.shots is not None:
+                results.append(
+                    m.process_samples(
+                        samples=self._samples,
+                        wire_order=self.wires,
+                        shot_range=shot_range,
+                        bin_size=bin_size,
+                    )
+                )
+
+            elif isinstance(m, StateMeasurement):
+                if self.shots is not None:
+                    warnings.warn(
+                        f"Requested measurement {m.__class__.__name__} with finite shots; the "
+                        "returned state information is analytic and is unaffected by sampling. "
+                        "To silence this warning, set shots=None on the device.",
+                        UserWarning,
+                    )
+                results.append(m.process_state(state=self.state, wire_order=self.wires))
+
+            elif isinstance(m, SampleMeasurement) and self.shots is None:
+                raise ValueError(
+                    "Shots must be specified in the device to compute the measurement "
+                    f"{m.__class__.__name__}"
+                )
```

Comment:
Can we group this with the other `SampleMeasurement` case above for better organization? Something like
```python
elif isinstance(m, SampleMeasurement):
    if self.shots is None:
        raise ValueError

    results.append(...)
```

Reply:
I refactored all this into a private function to make it clearer.